python/beeai_framework/cancellation.py

```diff
@@ -87,3 +93,25 @@ def trigger_abort(reason: str | None = None) -> None:
         if signal.aborted:
             trigger_abort(signal.reason)
         signal.add_event_listener(trigger_abort)
+
+
+async def signal_race(
+    fn: Callable[[], Awaitable[Any]], signal: AbortSignal | None = None, on_abort: Callable[[], None] | None = None
+) -> Awaitable[Any]:
+    def abort_handler() -> None:
+        if on_abort:
+            on_abort()
+        if signal:
+            signal.cancel()
+
+    if signal:
+        if signal.aborted:
+            raise AbortError(signal.reason)
+        else:
+            signal.add_event_listener(abort_handler)
+
+    try:
+        return await fn()
+    finally:
+        if signal:
+            signal.remove_event_listener(abort_handler)
```

Comment:
To have a real `race` you need to use `asyncio.as_completed([abort_signal, fn])` I would say.